enhancements/authentication/direct-external-oidc-provider.md

```diff
@@ -0,0 +1,316 @@
+
+### Drawbacks
+
+As mentioned above, configuring an external OIDC provider will effectively deactivate the built-in OAuth2 stack and remove all related API Services, resources and data. While switching back from an external OIDC provider to the built-in server is possible, it does not ensure that all existing data before the first switch to an OIDC provider will still exist, after reverting back to the built-in server.
+
+## Open Questions
+
+### Rollout strategy
+
+We must decide whether the updated KAS configuration will be rolled out via revisions or live config file reloading within the KAS pods.
+
+### console-operator
+
+- Does the console-operator need to wait for the KAS-o to report that OIDC is configured and available before proceeding with its reconfiguration? It currently only watches for auth type to be OIDC before reconfiguring. This might affect hypershift, as there is no KAS-o.
+
+### OAuth metadata & WebhookTokenAuthenticator cleanup
+
+Within the KAS-o, there is a [config observer](https://github.com/openshift/cluster-kube-apiserver-operator/blob/9fb047aaed45a81bf817233dde0ca09c8d38e257/pkg/operator/configobservation/auth/auth_metadata.go#L25) for the authentication CR which configures the value of `authConfig.oauthMetadataFile` of the `config` CM of the KAS, based on the `spec.oauthMetadata` or `status.integratedOAuthMetadata` values (the former takes precedence). If both are empty, the oauth-metadata configmap will be deleted.
+
+A similar [config observer](https://github.com/openshift/cluster-kube-apiserver-operator/blob/9fb047aaed45a81bf817233dde0ca09c8d38e257/pkg/operator/configobservation/auth/webhook_authenticator.go#L34) exists for the WebhookTokenAuthenticator as well, but this section discusses OAuth metadata only for simplicity; respective steps apply for the WebhookTokenAuthenticator as well.
+
+We can leverage this mechanism as-is for cleanup when configuring external OIDC. This could be done as follows:
+
+1. admin configures the authentication CR for OIDC
+2. this means that the `spec.oauthMetadata` field will be set to empty; however `status.integratedOAuthMetadata` will still have a value
+3. KAS-o config observer picks up the auth type change, configures KAS pods for OIDC, triggers a rollout
+4. once rollout is successful, CAO picks up the change to OIDC, removes oauth-specific resources, including `status.integratedOAuthMetadata`
+5. KAS-o oauth-metadata config observer picks up the change in `status.integratedOAuthMetadata` and since `spec.oauthMetadata` is also unset, removes the oauth-metadata configmap; this results in a new config, which triggers a new rollout
```

Comment:
mhm, I think that if you configure the external OIDC and remove the destination CM in the same step, no additional rollout will be required after cleaning up the status field, right?

Reply:
Actually, thinking more about it, I believe we cannot guarantee that these two operations will be performed atomically.

Reply:
What happens after configuring the external OIDC server? Does kas use the new one, or both?

Reply:
> if you configure the external OIDC and remove the destination CM in the same step, no additional rollout will be required after cleaning up the status field, right?
Right -- that's what I've tried to achieve with the alternative proposal. 

> we cannot guarantee that these two operations will be performed atomically
If I understand this correctly, we could achieve atomicity if we use the same config observer to configure either the webhook or OIDC depending on the auth type, in a mutually exclusive way, i.e. configuring the one will remove configuration of the other. Indeed we cannot guarantee this will happen atomically if we use different observers.

> Does kas use the new one, or both?
If we keep the webhook in the configuration and configure OIDC as well, both methods will be part of the authentication chain. If the first one in the chain fails to validate a token, then the next one will be used until it gets validated or the chain is finished. So both can be active at the same time.

One benefit of doing that, would be that we can keep the webhook up until the KAS-o has validated that OIDC works as expected; then it can mark its auth status as available, and the CAO can then clean up the oauth metadata and webhook (thus triggering a second rollout). If there's any problem with OIDC, the KAS-o won't mark OIDC as available, the CAO won't take down the webhook, and thus the webhook will keep working until the user fixes the OIDC config.

Reply:
Coming back to this for some further details I learned.

> If we keep the webhook in the configuration and configure OIDC as well, both methods will be part of the authentication chain.

While in theory this could be done, in practice that's not really possible due to our authentication CRD validation. The webhook can be defined only if type is _not_ OIDC, which means that if we switch configuration to OIDC in the auth CR then we have to get rid of the webhook.